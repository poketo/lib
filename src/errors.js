// @flow

type ErrorCode =
  | 'INVALID_URL'
  | 'UNSUPPORTED_SITE'
  | 'UNSUPPORTED_SITE_REQUEST'
  | 'TIMEOUT'
  | 'HTTP_ERROR';

class PoketoError extends Error {
  code: ErrorCode;
  constructor(errorCode: ErrorCode, message: string) {
    super(message);
    Error.captureStackTrace(this, this.constructor);
    this.name = this.constructor.name;
    this.code = errorCode;
  }
}

class InvalidUrlError extends PoketoError {
  constructor(url: string) {
    super('INVALID_URL', `Unable to parse url '${url}'`);
  }
}

class HTTPError extends PoketoError {
  statusCode: number;
  url: string;

  constructor(statusCode: number, message: string, url: string) {
    super('HTTP_ERROR', message);
    this.statusCode = statusCode;
    this.url = url;
  }
}

class NotFoundError extends HTTPError {
  constructor(url: string) {
    super(404, 'Not Found', url);
  }
}

class TimeoutError extends PoketoError {
  constructor(message: string, url: string) {
    super('TIMEOUT', message);
  }
}

class UnsupportedSiteError extends PoketoError {
  constructor(site: string) {
    super('UNSUPPORTED_SITE', `Site at '${site}' is not supported`);
  }
}

class UnsupportedSiteRequestError extends PoketoError {
  constructor(siteName: string, operationName: string) {
    super(
      'UNSUPPORTED_SITE_REQUEST',
      `${siteName} does not support ${operationName}`,
    );
  }
}

export default {
  PoketoError,
  HTTPError,
  InvalidUrlError,
  NotFoundError,
  TimeoutError,
  UnsupportedSiteError,
  UnsupportedSiteRequestError,
};
