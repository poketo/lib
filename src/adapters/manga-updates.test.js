import site from './manga-updates';
import errors from '../errors';

describe('MangaUpdatesAdapter', () => {
  const server = createVcrServer(site);

  beforeAll(async () => {
    const url = await server.listen(57161);
    site._getHost = () => url;
  });

  afterAll(() => {
    server.close();
  });

  describe('supportsUrl', () => {
    it('returns true for urls like mangaupdates.com', () => {
      expect(site.supportsUrl('http://mangaupdates.com')).toBe(true);
      expect(site.supportsUrl('https://www.mangaupdates.com')).toBe(true);
    });

    it('returns false for urls that are not mangaupdates.com', () => {
      expect(site.supportsUrl('http://www.mangahere.com')).toBe(false);
      expect(site.supportsUrl('http://ma.ngahere.co')).toBe(false);
    });
  });

  describe('parseUrl', () => {
    it('returns the components of a url', () => {
      expect(
        site.parseUrl('http://mangaupdates.com/series.html?id=111976'),
      ).toEqual({ seriesSlug: '111976', chapterSlug: null });
    });

    it('throws on unparseable urls', () => {
      expect(() => {
        site.parseUrl('https://www.mangaupdates.com/authors.html?id=462621');
      }).toThrow(errors.InvalidUrlError);
    });
  });

  describe('getSeries', () => {
    it('returns a metadata object', async () => {
      const metadata = await site.getSeries('111976');

      expect(metadata).toMatchSnapshot({
        updatedAt: expect.any(Number),
      });
    });
  });

  describe('getChapter', () => {
    it('throws an error', async () => {
      await expect(site.getChapter('111976')).rejects.toThrowError(
        errors.UnsupportedSiteRequest,
      );
    });
  });
});
