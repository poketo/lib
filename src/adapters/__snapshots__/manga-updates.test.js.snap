// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MangaUpdates getSeries returns a metadata object 1`] = `
Object {
  "artist": "Akitaka",
  "author": "Akitaka",
  "coverImageUrl": "https://www.mangaupdates.com/image/i276561.jpg",
  "description": "The Twelve City, a land governed by the leaders of the families who bear in their names the Chinese zodiac. In a city where youkai and humans coexist, the cat youkai who holds a grudge against the Rat Family appears in front of Nekuni Kyouichi, the head of the Rat Family. It's a malice-covered Hateful cat who is aiming for his life. But that youkai is Nekogasaki Natsuho, who's also extremely anxious around other people... Just like the saying \\"Curiosity killed the cat\\" goes, will these feelings also kill the cat?",
  "id": "manga-updates:111976",
  "publicationStatus": "ONGOING",
  "site": Object {
    "id": "manga-updates",
    "name": "MangaUpdates",
  },
  "slug": "111976",
  "supportsReading": false,
  "title": "Urami Koi, Koi, Urami Koi.",
  "updatedAt": Any<Number>,
  "url": "http://localhost:57161/series.html?id=111976",
}
`;
